model/common/src/icon4py/model/common/model_options.py

```diff
@@ -13,18 +13,43 @@
     Scalar,
     is_scalar_type,  # TODO(havogt): Should this function be public API?
 )
-from gt4py.next import Field, backend
+from gt4py.next import Field
 from gt4py.next.common import OffsetProvider
 from gt4py.next.ffront.decorator import Program
 
+from icon4py.model.common.model_backends import make_custom_dace_backend, make_custom_gtfn_backend
+
 
 def dict_values_to_list(d: dict[str, typing.Any]) -> dict[str, list]:
     return {k: [v] for k, v in d.items()}
 
 
+def get_options(program_name: str, arch: str, **backend):
+    backend_kind = backend.get("backend_kind", "default")
+    return dict(backend_kind=backend_kind)
+
+
+def customize_backend(program: str = "", arch: str = "", **backend):
+    program_id = str(program.past_stage.past_node.id) if program != "" else program
+    options = get_options(program_id, arch, **backend)
+    if options["backend_kind"] == "dace":
+        backend_func = make_custom_dace_backend
+    elif options["backend_kind"] == "gtfn":
+        backend_func = make_custom_gtfn_backend
+    on_gpu = backend["device"] == "gpu"
+    custom_backend = backend_func(
+        on_gpu=on_gpu,
+        **options,
+    )
+    return custom_backend
```

Comment:
A few more thoughts (I make one comment per idea, to be able to discuss them separately)

Reply:
Maybe instead of `backend_kind` we could use `backend_factory` directly, i.e. the advanced user (aka performance engineer) instead of passing a str-identifier could directly pass the `make_custom_dace_backend` function.
It would look like:
```suggestion
def customize_backend(program: str = "", arch: str = "", **backend):
    program_id = str(program.past_stage.past_node.id) if program != "" else program
    options = get_options(program_id, arch, **backend)
    on_gpu = backend["device"] == "gpu"
    backend_factory = options.pop("backend_factory")
    custom_backend = backend_factory(
        on_gpu=on_gpu,
        **options,
    )
    return custom_backend
```
and
```python
def get_options(program_name: str, arch: str, **backend):
    backend_factory = backend.get("backend_factory", make_custom_dace_backend)
    return dict(backend_kind=backend_kind)
```

Reply:
where would you switch between `make_custom_dace_backend` and `make_custom_gtfn_backend`

Reply:
The user would pass one or the other and if nothing is passed the default will be used.

Reply:
I don't think that this is a user-friendly approach. They should be able to just say "gtfn" or "dace". We can make gtfn`backend_kind` 's default

Reply:
If it's just about the name you can make an alias from `gtfn` to `make_custom_gtfn_backend`.

Reply:
That makes sense to me. Especially, with the last statement. 
>If that was me, I would want to only specify "CPU" or "GPU" and then rely on the others to run whatever backend and backend customization is fastest for the given architecture.

Some observations:
- In this PR, so far we are discussing the granule interface, how we map the driver command line arguments was only partially touched in the discussion, but I agree it makes sense to think about that as well.
- passing a concrete backend to the driver might become a corner-case, because even if the user (= the one who runs the application) wants to say `dace` + `gpu`, they probably would like to run with the best dace variant for every gt4py program, not with the concrete backend that is constructed in icon4py (or gt4py).
- passing a factory can be implemented for the command line similar to how we added passing custom backends to the stencil_tests, there we allow the pattern `--backend my_package.subpackage:my_backend` (see also https://docs.python.org/3/library/pkgutil.html#pkgutil.resolve_name for the standard way to load an object from such a string). Additionally we could make one package `icon4py.model.common.model_backends` or similar the default backend, i.e. if the string doesn't contain a `:`, we could prefix with that package. We could than do `driver --device=gpu, --backend-factory=dace`, which would pass `icon4py.model.common.model_backends.make_custom_dace_backend` (after making an alias `dace` -> `make_custom_dace_backend`, but these are details)